### PR TITLE
Update markbind-action to enhance existing features

### PR DESCRIPTION
**Pull request by tlylt**
Details of this PR has been posted & discussed in the main repo [here](https://github.com/MarkBind/markbind/pull/1839).

Proposed commit message:
Update markbind-action to enhance existing features

The prerelease markbind-action has certain limitations such
as cannot specify the version of MarkBind to build files or 
prevent force-pushing when updating the gh-pages branch.

Let's update it to a composite action and expand on the
existing functionalities to make this action more useful.
New configurations have been included to allow for flexibility
and support more use cases.

Using a composite action also makes it easier to maintain and
reuse existing actions.

### CODE REVIEW

File: README.md
Comment (ang-zeyu) [inline README.md]: does this support semantic versioning? might be good to add a note if so

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
  - This will preserve any existing files in the published branch before an update is made.
```

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
## Behind the Scenes
```

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
Besides processing a few bash scripts, this action utilises the following actions:
```

File: README.md
Comment (ang-zeyu) [inline README.md]: would `'development'` be more apt?

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
  - This is the latest, possibly unpublished version of MarkBind in development
```

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
  - Note that for PR Preview purposes, the domain you specify will be prefixed with 'pr-x-', where 'x' is the GitHub event number
```

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
- 'pr-x-<domain>'
```

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion

```

File: README.md
Comment (ang-zeyu) [inline README.md]: Let's move this note up to the `domain` option. Perhaps we could also add a note on setting up subdomains for pr previews (I'm assuming, not too familiar with surge).

**Review by ang-zeyu — COMMENTED** [review]
Nice work, especially on the recipes.

Formatting wise, should we enclose workflow variables with double backticks for clarity? (e.g. `'gh-pages'`) Not a huge issue though, if the current looks nicer.

Just a few nits otherwise, mostly on the README:

File: README.md
Comment (ang-zeyu) [inline README.md]: One more question / suggestion: Is this action able to automatically cleanup merged PR deployments? If not, could also add a reminder on this to manually do so.

File: README.md
Comment (tlylt) [reply README.md]: Yup will add that.

According to [this](https://docs.npmjs.com/cli/v6/commands/npm-install) (internally installing via npm install) it should support version ranges. I tried doing something like `npm install cross-env@^7.0.0` or `npm install cross-env@~7.0.0` and both seem to work as well so I guess semantic versioning is also fine.

File: README.md
Comment (tlylt) [reply README.md]: For PR deployments (via Surge), the deployment will not be automatically deleted. This may be possible but the user will need to create another action to trigger the deletion on PR merge/close following this [instruction](https://surge.sh/help/tearing-down-a-project). I guess I can create that as a reusable workflow and link the usage here so that users can choose to include that additional step if needed.

File: README.md
Comment (ang-zeyu) [inline README.md]: Let's add a note that this is for `service=surge` only as well

File: README.md
Comment (ang-zeyu) [inline README.md]:
```suggestion
The domain that the site is available at. Required if service chosen is Surge.
```
(relevant for both pr preview / deployment if i'm not mistaken)

File: README.md
Comment (ang-zeyu) [inline README.md]: same edits here for `purpose / domain`

**Review by ang-zeyu — COMMENTED** [review]
> For PR deployments (via Surge), the deployment will not be automatically deleted. This may be possible but the user will need to create another action to trigger the deletion on PR merge/close following this [instruction](https://surge.sh/help/tearing-down-a-project). I guess I can create that as a reusable workflow and link the usage here so that users can choose to include that additional step if needed.

Sounds good, let's put up an issue for tracking. (if not attempting anytime soon)

Last few nits:

File: README.md
Comment (jonahtanjz) [inline README.md]: Really minor suggestion: Would it be better to make the `Option`s link to their respective `Option Details` section? I think this can help the users to quickly look at the detailed description to get the possible values etc.

**Review by jonahtanjz — COMMENTED** [review]
Nice work @tlylt!

Overall it looks good 👍  Just one suggestion :)

**Comment by tlylt** [comment]
@ang-zeyu @jonahtanjz Thanks for checking!

I think I have updated all the required stuff (let me know if there are any more issues).

For the release plan, according to [this guide for Action release management](https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-tags-for-release-management), I think the steps will be:
- Fix any last issues within this PR
- Merge this into master
- Create a `release/v2` branch from master for testing
- When ready, tag master as 'v2.0.0', and also 'v2', release on GitHub
- The 'v2' tag will be shifted everytime there's a new release of v2 level version

**Review by jonahtanjz — APPROVED** [review]
LGTM 👍 

Let's wait for @ang-zeyu to approve as well :)

**Review by ang-zeyu — APPROVED** [review]

**Comment by ang-zeyu** [comment]
Lgtm 👍 

Thanks for linking the guide as well!

**Comment by tlylt** [comment]
Seems like there's a regression possibly to some updates that I did in this PR such that the `baseUrl` input is not being handled properly if it is `""`. I will update here again once the issue is resolved.

**Comment by tlylt** [comment]
Sorry for the delay. I retested the three cases of baseUrl input and they all seem to work now. I think we are good to proceed with 
> Merge this into master
> Create a release/v2 branch from master for testing

**Comment by jonahtanjz** [comment]
@tlylt Do you have write access to do the tagging and release?

Might also want to include a short commit message in the PR description :)

**Comment by tlylt** [comment]
> @tlylt Do you have write access to do the tagging and release?
> 
> Might also want to include a short commit message in the PR description :)

I have added a commit message. I think I may not have the right to do so (I also don't think I can merge this PR).

**Comment by ang-zeyu** [comment]
@tlylt you should have permissions now : )

**Pull request merged**
